horovod/common/operations.cc

```diff
@@ -272,14 +121,40 @@ const Status DUPLICATE_NAME_ERROR = Status::InvalidArgument(
     "name as another tensor that is currently being processed.  If you want "
     "to request another tensor, use a different tensor name.");
 
-#define OP_ERROR(entries, error_message)                                       \
-  {                                                                            \
-    for (auto& e : (entries)) {                                                \
-      timeline.End(e.tensor_name, nullptr);                                    \
-      e.callback(Status::UnknownError(error_message));                         \
-    }                                                                          \
-    return;                                                                    \
-  }
+OperationManager* CreateOperationManager(HorovodGlobalState& state) {
+  // Order of these operations is very important. Operations will be checked sequentially from the first
```

Comment:
Would it make sense to have an explicit decision function to choose an implementation for each op dynamically? I feel like the sequential ordering makes the decision a bit implicit, and might make it harder to write more complex logic later, for example dynamically choosing an algorithm on a per-tensor basis.

Reply:
Personally, I think the sequencing approach reduces the complexity (but perhaps I'm biased from writing so many behavior trees over the years).

There's some complication that arises from using an explicit rule, which is that some of the ops available at runtime will be dynamic (based on what frameworks you have installed: NCCL, DDL, etc.), while others will be available given some parameter state, and still others (as you mention) will be available on a per-tensor basis (which is available to the Enabled function via the Response parameter).

Using an explicit rule, the operation manager becomes very complicated with preprocessor code, needing to couple itself to what frameworks are installed (because certain ops will not exist at compile time).  Using the sequential approach, all of that is isolated to a single function in `operations.cc`, which I think makes the separation between "what ops are available at compile time" and "what ops are available at runtime" more clear.

Reply:
Makes sense. Would it be easy to override the decision logic implied by this list if needed? Is it sufficient to call `execute()` for the desired op in `OperationManager::ExecuteAllreduce`, or would it be more complicated?

Reply:
Yes, it should be a pretty simple change to enforce the use a certain operation under certain conditions.  It would just require giving the `OperationManager` a direct pointer to the operation or having a map (right now, we don't directly reference the operations, only store them in the list).